src/Kinto.elm

```diff
@@ -0,0 +1,316 @@
+                joinUrl [ baseUrl, "buckets", bucketName, "collections", collectionName, "records", recordId ]
+
+
+client : Config -> Endpoint -> Verb -> Task Error Json.Value
+client config endpoint verb =
+    let
+        request =
+            { verb = verb
+            , url = endpointUrl config endpoint
+            , headers = config.headers
+            , body = Http.empty
+            }
+    in
+        (Http.send Http.defaultSettings request)
+            |> toKintoResponse
+
+
+withHeader : String -> String -> Config -> Config
+withHeader name value ({ headers } as config) =
+    let
+        newHeaders =
+            ( name, value ) :: headers
+    in
+        { config | headers = newHeaders }
+
+
+withAuthHeader : Auth -> Config -> Result Error Config
+withAuthHeader auth config =
+    case auth of
+        NoAuth ->
+            Ok config
+
+        Basic username password ->
+            let
+                hash str =
+                    Base64.encode str |> Result.formatError ConfigError
+
+                addAuthHeader hash =
+                    Ok (withHeader "Authorization" ("Basic " ++ hash) config)
+            in
+                hash (username ++ ":" ++ password)
+                    `Result.andThen` addAuthHeader
+
+        Bearer token ->
+            Ok (config |> withHeader "Authorization" ("Bearer " ++ token))
+
+
+configure : Url -> Auth -> Result Error Config
+configure baseUrl auth =
+    Config baseUrl []
+        |> withAuthHeader auth
+        |> Debug.log "Kinto request"
+
+
+
+-- Dealing with answers from the Kinto server
+
+
+bodyDataDecoder : Json.Decoder Json.Value
+bodyDataDecoder =
+    ("data" := Json.value)
+
+
+extractData : String -> StatusCode -> StatusMsg -> Task Error Json.Value
+extractData body statusCode statusMsg =
+    let
+        responseResult =
+            Json.decodeString bodyDataDecoder body
+    in
+        case responseResult of
+            Ok data ->
+                Task.succeed data
+
+            Err _ ->
+                -- Is it an ErrorRecord json?
+                decodeError body statusCode statusMsg
+
+
+errorDecoder : Json.Decoder ErrorRecord
+errorDecoder =
+    Json.object4 ErrorRecord
+        ("errno" := Json.int)
+        ("message" := Json.string)
+        ("code" := Json.int)
+        ("error" := Json.string)
+
+
+decodeError : String -> StatusCode -> StatusMsg -> Task Error a
+decodeError body statusCode statusMsg =
+    case Json.decodeString errorDecoder body of
+        Ok errorRecord ->
+            Task.fail (KintoError statusCode statusMsg errorRecord)
+
+        Err msg ->
+            Task.fail (ServerError statusCode statusMsg body)
+
+
+toKintoResponse : Task Http.RawError Http.Response -> Task Error Json.Value
+toKintoResponse response =
+    (Task.mapError NetworkError response)
+        `Task.andThen` (handleResponse extractData)
+
+
+handleResponse :
+    (String -> StatusCode -> StatusMsg -> Task Error Json.Value)
```

Comment:
Yes, this is almost exactly what I was thinking.

I might change `extractData` to produce a `Result Error foo` instead of a `Result String foo`, but I don't think it matters much.

To restate the advantages here:

- We can glance at the type signatures and because we don't see `Task`, we know the function is pure. In general, the type signatures align more with the underlying functions -- `Result foo bar` meaning "pure function that can fail".
- `toKintoResponse` can be called on any `Response`, whether it came from a `Task` or not. This is another way where it's clearer what `toKintoResponse` actually uses as its input.
- You can still use `extractBody` with different decoders, but it's still relatively simple to read and use because it isn't a higher-order function.

One thing that may be a disadvantage:
- `performQuery` is now itself responsible for managing the error type from the request.